File: solutions/zig/collatz-conjecture/1/collatz_conjecture.zig
pub const ComputationError = error{IllegalArgument};

pub fn steps(number: usize) anyerror!usize {
    if (number <= 0) return ComputationError.IllegalArgument;

    var s: usize = 0;
    var n = number;

    while (n > 1) : (s += 1) {
        if (n % 2 == 0) {
            n /= 2;
        } else {
            n = 3 * n + 1;
        }
    }
    return s;
}

File: solutions/swift/collatz-conjecture/2/Sources/CollatzConjecture/CollatzConjecture.swift
enum CollatzConjectureError: Error { case negative }

class CollatzConjecture {
  static func steps(_ number: Int) throws -> Int {
    guard number > 0 else { throw CollatzConjectureError.negative }
    return number == 1
      ? 0
      : try steps(number % 2 == 0 ? number / 2 : number * 3 + 1) + 1
  }
}
